packages/ai-jsx/src/lib/openai.tsx

```diff
@@ -438,59 +422,97 @@ export async function* OpenAIChatModel(
     }
   >;
 
-  const currentMessage = { content: undefined, function_call: undefined } as Partial<ChatCompletionResponseMessage>;
-  let finishReason: string | undefined = undefined;
-  for await (const deltaMessage of openAiEventsToJson<ChatCompletionDelta>(
-    asyncIteratorOfFetchStream(chatResponse.body!.getReader())
-  )) {
-    logger.trace({ deltaMessage }, 'Got delta message');
-    finishReason = finishReason ?? deltaMessage.choices[0].finish_reason;
-    const delta = deltaMessage.choices[0].delta;
-    if (delta.role) {
-      currentMessage.role = deltaMessage.choices[0].delta.role;
+  const iterator = openAiEventsToJson<ChatCompletionDelta>(asyncIteratorOfFetchStream(chatResponse.body!.getReader()))[
+    Symbol.asyncIterator
+  ]();
+
+  // We have a single response iterator, but we'll wrap tokens _within_ the structure of <AssistantMessage> or <FunctionCall>
+  // components. This:
+  //  - Allows our stream to be append-only and therefore eagerly rendered in append-only contexts.
+  //  - Preserves the output structure to allow callers to extract/separate <AssistantMessage> and <FunctionCall> messages.
+  //  - Allows the intermediate states of the stream to include "partial" <FunctionCall> elements with healed JSON.
+  //
+  // This requires some gymnastics because several components will share a single iterator that can only be consumed once.
+  // That is, the logical loop execution is spread over multiple functions (closures over the shared iterator).
+  async function advance(): Promise<Partial<ChatCompletionResponseMessage> | null> {
+    const next = await iterator.next();
+    if (next.done) {
+      return null;
     }
-    if (delta.content) {
-      currentMessage.content = currentMessage.content ?? '';
-      currentMessage.content += delta.content;
-      if (!props.experimental_streamFunctionCallOnly) {
-        yield delta.content;
-      }
+
+    logger.trace({ deltaMessage: next.value }, 'Got delta message');
+    return next.value.choices[0].delta;
+  }
+
+  let delta = await advance();
+  while (delta !== null) {
+    if (delta.role === 'assistant') {
+      // Memoize the stream to ensure it renders only once.
+      const assistantStream = memo(
+        (async function* (): AI.RenderableStream {
+          yield AI.AppendOnlyStream;
+
+          while (delta !== null) {
+            if (delta.content) {
+              yield delta.content;
+            }
+            if (delta.function_call) {
+              break;
+            }
+            delta = await advance();
+          }
+
+          return AI.AppendOnlyStream;
+        })()
+      );
+      yield <AssistantMessage>{assistantStream}</AssistantMessage>;
+
+      // Ensure the assistantStream() is flushed by rendering it.
+      await render(assistantStream);
```

Comment:
Q: since this is a `memo` object, rendering it (without checking the output) is enough?
Otherwise this piece of code is confusing.

Reply:
Yeah it's more than a little sketchy:

- `memo` ensures that side-effects of the generator will be consumed _at most_ once
- `yield`ing the generator inside a component does not guarantee that the generator will be executed (it will only be executed if the yielded value is actually rendered)
- `render`ing the memoized generator guarantees that the generator will be consumed _at least_ once